src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,13 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Base class for all valid values generators with cache support.
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.
+        /// By default 'ValidValuesCacheExpiration = 0' and no valid values is cached (is re-evaluated every time it is used).
+        /// </summary>
+        public int ValidValuesCacheExpiration { 
```

Comment:
I think this class should have a single constructor:
```C#
public CachedValidValuesGeneratorBase(int cacheExpirationInSeconds) { _validValuesCacheExpiration = cacheExpirationInSeconds; }
```

I don't think the property should be exposed - as changing the value implies something (e.g. that I can extend or shorten the cache expiration for values already cached.)